### DISCUSSION TITLE
When performing validation is it possible to preserve the errors array?

### DISCUSSION THREAD

**Discussion by TomppaPackage** [TanStack/form#860] [Q&A] [answered]
When using `@tanstack/zod-form-adapter` & ZodAdapter with Zod for Field-Level validation, the `field.state.meta.errors` array is concatenated into a single string.

Is it possible to preserve the original error array?

For instance, given the following schema;
```
const fullNameSchema = z
  .string()
  .min(12, { message: "Full name must contain at least 12 characters" })
  .refine((value) => value.startsWith("a"), {
    message: 'Full name must start with "a"',
  })
  .refine((value) => value.endsWith("b"), {
    message: 'Full name must end with "b"',
  })
  .refine((value) => value.includes("c"), {
    message: 'Full name must contain "c"',
  });
```

The output of `fullNameSchema.safeParse('HelloWorld')` will look like this;
```
[
    {
        "code": "too_small",
        "minimum": 12,
        "type": "string",
        "inclusive": true,
        "exact": false,
        "message": "Full name must contain at least 12 characters",
        "path": []
    },
    {
        "code": "custom",
        "message": "Full name must start with \"a\"",
        "path": []
    },
    {
        "code": "custom",
        "message": "Full name must end with \"b\"",
        "path": []
    },
    {
        "code": "custom",
        "message": "Full name must contain \"c\"",
        "path": []
    }
]
```

However, if using ZodAdapter, the output would be;

```
[
  'Full name must contain at least 12 characters, Full name must start with "a", Full name must end with "b", Full name must contain "c"'
]
```

Link to a reproducible example.
https://codesandbox.io/p/sandbox/cool-andras-2vjr3c?file=%2Fsrc%2Findex.tsx

**Comment by my-slab** [comment, marked as answer]
I don't think so, because a validation can only return a `ValidationError` which is a `string`. However, you can access the issues array and transform the error message to your liking using the `transformErrors` function that is passed to the adapter.

e.g., just show one of the error messages.

```tsx
 validatorAdapter: zodValidator({
    transformErrors: (e) => {
      return e?.[0]?.message
    },
  }),
```

**Reply by TomppaPackage** [reply]
It's true we can transform the resulting text, but in our use-case we present errors (when multiple) as a list. 

I've found a workaround for the time being, which is to just add a clear separator in the `transformErrors` fn, and then split the resulting string by that separator to get an array of error strings.

This does however answer my question, that it's not a use-case currently supported by `@tanstack/form`, so I will mark this as answered. Thank you! :)
